### ISSUE TITLE
Use pureconfig for CouchDbRestStore

### ISSUE THREAD

**Issue by chetanmeh** [apache/incubator-openwhisk#3355]
Currently required properties related to ArtifactStore is mentioned at various places like.

1. WhiskAuthStore
2. WhiskEntityStore
3. WhiskActivationStore

The properties required are

```
  def requiredProperties =
    Map(
      dbProvider -> null,
      dbProtocol -> null,
      dbUsername -> null,
      dbPassword -> null,
      dbHost -> null,
      dbPort -> null,
      dbAuths -> null)
```

These properties are specific to CouchDB implementation and may not be required for a different ArtifactStore implementation

We should add  a `requiredProperties` to `ArtifactStore` (like done in `LoadBalancerProvider`) such that the implementation can govern which all properties are required

**Comment by markusthoemmes** [comment]
Alternatively, we should move the ArtifactStore to `pureconfig` which completely localizes the concern to the implementation.

**Comment by chetanmeh** [comment]
> we should move the ArtifactStore to pureconfig

Eventually yes but I believe such a change would touch multiple places (ansible scripts) as I am not sure it can be done in backward compatible way. Or can it be done in backward compatible way?

**Comment by markusthoemmes** [comment]
In theory it can but in the past we did all the pureconfig moves by "biting the bullet" and fix every deployment method lazily.

Only the direct value->container mapping is impacted (controller's and invoker's deploy.yml in this case)

**Comment by chetanmeh** [comment]
Okie would then work on PR which modifies the CouchDBStore to use pureconfig and see if it can be done in backward compatible way say by using substitution as all the current properties are passed via environment variables

**Comment by chetanmeh** [comment]
Closed via apache/incubator-openwhisk#3369
